### DISCUSSION TITLE
关于SM9 非XOR加密标准问题

### DISCUSSION THREAD

**Discussion by emmansun** [emmansun/gmsm#112] [General]
SM9非XOR加密，有几个模式是需要IV的，SM9标准对这一部分没有描述。关于IV的处理，有两个思路：

1. 目前的实现方式，随机生成，附加在密文头部；
2. 通过密钥扩展函数和加密密钥一起生成，没有传输的问题；但是和确定密钥和IV的顺序。

两种方式都可实现，取决于标准定义，要不然无法互操作。

**Comment by lyc666-beep** [comment]
可不可以用cmac

**Comment by lyc666-beep** [comment]
![新建 BMP 图像](https://github.com/user-attachments/assets/394ad2f4-e960-465e-9f78-717920f7ea4c)
可不可以通过这个确定到底用的是什么算法

**Comment by lyc666-beep** [comment]
gmt0044.5

**Comment by emmansun** [comment]
不是很明白你的意思。本项目已经实现当前规范列出的模式:
```go
const (
	ENC_TYPE_XOR encryptType = 0
	ENC_TYPE_ECB encryptType = 1
	ENC_TYPE_CBC encryptType = 2
	ENC_TYPE_OFB encryptType = 4
	ENC_TYPE_CFB encryptType = 8
)
```
我这里的意思是，有些模式可能互操作有问题，毕竟没有说明如IV的生成和传输等规范。

**Comment by lyc666-beep** [comment]
是我理解错了，我以为是说用那个算法



---原始邮件---
发件人: "Sun ***@***.***&gt;
发送时间: 2024年11月21日(周四) 晚上10:01
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [emmansun/gmsm] 关于SM9 非XOR加密标准问题 (Discussion #112)




 
不是很明白你的意思。本项目已经实现当前规范列出的模式:
 const ( 	ENC_TYPE_XOR encryptType = 0 	ENC_TYPE_ECB encryptType = 1 	ENC_TYPE_CBC encryptType = 2 	ENC_TYPE_OFB encryptType = 4 	ENC_TYPE_CFB encryptType = 8 )
 
我这里的意思是，有些模式可能互操作有问题，毕竟没有说明如IV的生成和传输等规范。

**Comment by HChen4811** [comment]
按GM/T 0018-2023的接口定义，SM9加解密中涉及的IV是和SM9密文不相关的参数，需要独立于SM9密文结构，IV是独立于SM9加解密过程的参数而不应由SM9加解密过程自行产生

**Reply by emmansun** [reply]
GM/T 0018-2023是《密码设备应用接口规范》，IV的话，涉及产生和传输。

- 产生：目前实现通过传入`io.Reader`接口实现，调用者可以使用`rand.Reader`，或自定义实现，比如通过密码硬件；
- 传输：象PKCS有关标准的话，都会用语义化的ASN1定义。目前实现只是简单拼接。

**Comment by lyc666-beep** [comment]
好的谢谢



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: HChen4811 ***@***.***&gt;
发送时间: 2025年1月14日 10:32
收件人: emmansun/gmsm ***@***.***&gt;
抄送: lyc666-beep ***@***.***&gt;, Comment ***@***.***&gt;
主题: Re: [emmansun/gmsm] 关于SM9 非XOR加密标准问题 (Discussion #112)





 
按GM/T 0018-2023的接口定义，SM9加解密中涉及的IV是和SM9密文不相关的参数，需要独立于SM9密文结构，IV是独立于SM9加解密过程的参数而不应由SM9加解密过程自行产生

**Comment by lyc666-beep** [comment]
好的谢谢



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: Sun Yimin ***@***.***&gt;
发送时间: 2025年1月14日 11:42
收件人: emmansun/gmsm ***@***.***&gt;
抄送: lyc666-beep ***@***.***&gt;, Comment ***@***.***&gt;
主题: Re: [emmansun/gmsm] 关于SM9 非XOR加密标准问题 (Discussion #112)





 
GM/T 0018-2023是《密码设备应用接口规范》，IV的话，涉及产生和传输。
  
产生：目前实现通过传入io.Reader接口实现，调用者可以使用rand.Reader，或自定义实现，比如通过密码硬件；
 
传输：象PKCS有关标准的话，都会用语义化的ASN1定义。目前实现只是简单拼接。
